### PR TITLE
Improve SSE2 slide hash performance

### PR DESCRIPTION
**Pull request by KungFuJesus**
At least on pre-nehalem CPUs, we get a > 50% improvement. This is
mostly due to the fact that we're opportunistically doing aligned loads
instead of unaligned loads.  This is something that is very likely to be
possible, given that the deflate stream initialization uses the zalloc
function, which most libraries don't override.  Our allocator aligns to
64 byte boundaries, meaning we can do aligned loads on even AVX512 for
the zstream->prev and zstream->head pointers. However, only pre-nehalem
CPUs _actually_ benefit from explicitly aligned load instructions.

The other thing being done here is we're unrolling the loop by a factor
of 2 so that we can get a tiny bit more ILP.  This improved performance
by another 5%-7% gain.

### CODE REVIEW

**Comment by KungFuJesus** [comment]
The improvement as measured on a Q9650:

```
Comparing slide_hash/sse2 (from ./benchmark_zlib_old) to slide_hash/sse2 (from ./benchmark_zlib)
Benchmark                                                            Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------------
[slide_hash/sse2 vs. slide_hash/sse2]/1024                        -0.5738         -0.5739         16804          7162         16804          7160
[slide_hash/sse2 vs. slide_hash/sse2]/1024                        -0.5728         -0.5729         16790          7172         16790          7171
[slide_hash/sse2 vs. slide_hash/sse2]/1024                        -0.5725         -0.5727         16778          7172         16777          7170
[slide_hash/sse2 vs. slide_hash/sse2]/1024                        -0.5733         -0.5733         16805          7170         16805          7170
[slide_hash/sse2 vs. slide_hash/sse2]/1024                        -0.5736         -0.5736         16787          7158         16787          7158
[slide_hash/sse2 vs. slide_hash/sse2]/1024                        -0.5740         -0.5740         16798          7156         16797          7156
[slide_hash/sse2 vs. slide_hash/sse2]/1024                        -0.5731         -0.5730         16772          7160         16767          7160
[slide_hash/sse2 vs. slide_hash/sse2]/1024                        -0.5726         -0.5724         16760          7164         16753          7164
[slide_hash/sse2 vs. slide_hash/sse2]/1024                        -0.5712         -0.5712         16745          7180         16744          7180
[slide_hash/sse2 vs. slide_hash/sse2]/1024                        -0.5727         -0.5726         16746          7156         16743          7156
[slide_hash/sse2 vs. slide_hash/sse2]/1024_pvalue                  0.0002          0.0002      U Test, Repetitions: 10 vs 10
[slide_hash/sse2 vs. slide_hash/sse2]/1024_mean                   -0.5730         -0.5730         16778          7165         16777          7164
[slide_hash/sse2 vs. slide_hash/sse2]/1024_median                 -0.5732         -0.5732         16782          7163         16782          7162
[slide_hash/sse2 vs. slide_hash/sse2]/1024_stddev                 -0.6311         -0.6594            22             8            24             8
[slide_hash/sse2 vs. slide_hash/sse2]/1024_cv                     -0.1362         -0.2025             0             0             0             0
[slide_hash/sse2 vs. slide_hash/sse2]/2048                        -0.5717         -0.5717         17013          7286         17013          7286
[slide_hash/sse2 vs. slide_hash/sse2]/2048                        -0.5713         -0.5712         17021          7296         17015          7296
[slide_hash/sse2 vs. slide_hash/sse2]/2048                        -0.5716         -0.5716         17024          7293         17024          7293
[slide_hash/sse2 vs. slide_hash/sse2]/2048                        -0.5716         -0.5716         17030          7295         17030          7295
[slide_hash/sse2 vs. slide_hash/sse2]/2048                        -0.5731         -0.5732         17057          7281         17057          7280
[slide_hash/sse2 vs. slide_hash/sse2]/2048                        -0.5726         -0.5726         17017          7273         17017          7273
[slide_hash/sse2 vs. slide_hash/sse2]/2048                        -0.5725         -0.5725         17040          7284         17040          7284
[slide_hash/sse2 vs. slide_hash/sse2]/2048                        -0.5717         -0.5718         17010          7285         17009          7284
[slide_hash/sse2 vs. slide_hash/sse2]/2048                        -0.5732         -0.5732         17048          7277         17048          7275
[slide_hash/sse2 vs. slide_hash/sse2]/2048                        -0.5721         -0.5721         17001          7275         17001          7275
[slide_hash/sse2 vs. slide_hash/sse2]/2048_pvalue                  0.0002          0.0002      U Test, Repetitions: 10 vs 10
[slide_hash/sse2 vs. slide_hash/sse2]/2048_mean                   -0.5722         -0.5722         17026          7285         17025          7284
[slide_hash/sse2 vs. slide_hash/sse2]/2048_median                 -0.5721         -0.5720         17023          7284         17020          7284
[slide_hash/sse2 vs. slide_hash/sse2]/2048_stddev                 -0.5339         -0.5315            18             8            18             8
[slide_hash/sse2 vs. slide_hash/sse2]/2048_cv                     +0.0893         +0.0951             0             0             0             0
[slide_hash/sse2 vs. slide_hash/sse2]/4096                        -0.5728         -0.5728         17548          7496         17544          7495
[slide_hash/sse2 vs. slide_hash/sse2]/4096                        -0.5723         -0.5725         17543          7502         17542          7499
[slide_hash/sse2 vs. slide_hash/sse2]/4096                        -0.5744         -0.5743         17596          7490         17592          7489
[slide_hash/sse2 vs. slide_hash/sse2]/4096                        -0.5742         -0.5743         17591          7490         17591          7488
[slide_hash/sse2 vs. slide_hash/sse2]/4096                        -0.5735         -0.5736         17565          7491         17565          7490
[slide_hash/sse2 vs. slide_hash/sse2]/4096                        -0.5740         -0.5740         17578          7489         17578          7489
[slide_hash/sse2 vs. slide_hash/sse2]/4096                        -0.5734         -0.5734         17558          7491         17558          7491
[slide_hash/sse2 vs. slide_hash/sse2]/4096                        -0.5738         -0.5738         17568          7488         17568          7488
[slide_hash/sse2 vs. slide_hash/sse2]/4096                        -0.5744         -0.5744         17593          7487         17593          7487
[slide_hash/sse2 vs. slide_hash/sse2]/4096                        -0.5742         -0.5742         17580          7485         17580          7485
[slide_hash/sse2 vs. slide_hash/sse2]/4096_pvalue                  0.0002          0.0002      U Test, Repetitions: 10 vs 10
[slide_hash/sse2 vs. slide_hash/sse2]/4096_mean                   -0.5737         -0.5737         17572          7491         17571          7490
[slide_hash/sse2 vs. slide_hash/sse2]/4096_median                 -0.5738         -0.5738         17573          7490         17573          7489
[slide_hash/sse2 vs. slide_hash/sse2]/4096_stddev                 -0.7438         -0.7829            19             5            19             4
[slide_hash/sse2 vs. slide_hash/sse2]/4096_cv                     -0.3989         -0.4906             0             0             0             0
[slide_hash/sse2 vs. slide_hash/sse2]/8192                        -0.5721         -0.5722         18579          7950         18579          7948
[slide_hash/sse2 vs. slide_hash/sse2]/8192                        -0.5724         -0.5724         18589          7948         18586          7947
[slide_hash/sse2 vs. slide_hash/sse2]/8192                        -0.5726         -0.5727         18605          7952         18604          7950
[slide_hash/sse2 vs. slide_hash/sse2]/8192                        -0.5732         -0.5732         18603          7939         18600          7939
[slide_hash/sse2 vs. slide_hash/sse2]/8192                        -0.5728         -0.5728         18571          7933         18570          7933
[slide_hash/sse2 vs. slide_hash/sse2]/8192                        -0.5731         -0.5731         18564          7925         18561          7925
[slide_hash/sse2 vs. slide_hash/sse2]/8192                        -0.5722         -0.5722         18550          7936         18550          7935
[slide_hash/sse2 vs. slide_hash/sse2]/8192                        -0.5732         -0.5732         18586          7932         18585          7932
[slide_hash/sse2 vs. slide_hash/sse2]/8192                        -0.5715         -0.5716         18548          7948         18548          7946
[slide_hash/sse2 vs. slide_hash/sse2]/8192                        -0.5721         -0.5721         18556          7941         18556          7940
[slide_hash/sse2 vs. slide_hash/sse2]/8192_pvalue                  0.0002          0.0002      U Test, Repetitions: 10 vs 10
[slide_hash/sse2 vs. slide_hash/sse2]/8192_mean                   -0.5725         -0.5725         18575          7940         18574          7940
[slide_hash/sse2 vs. slide_hash/sse2]/8192_median                 -0.5725         -0.5726         18575          7940         18574          7939
[slide_hash/sse2 vs. slide_hash/sse2]/8192_stddev                 -0.5618         -0.5797            21             9            20             8
[slide_hash/sse2 vs. slide_hash/sse2]/8192_cv                     +0.0251         -0.0167             0             0             0             0
[slide_hash/sse2 vs. slide_hash/sse2]/16384                       -0.5729         -0.5728         20676          8831         20667          8829
[slide_hash/sse2 vs. slide_hash/sse2]/16384                       -0.5736         -0.5736         20652          8806         20651          8806
[slide_hash/sse2 vs. slide_hash/sse2]/16384                       -0.5731         -0.5731         20658          8818         20654          8818
[slide_hash/sse2 vs. slide_hash/sse2]/16384                       -0.5724         -0.5724         20657          8832         20656          8832
[slide_hash/sse2 vs. slide_hash/sse2]/16384                       -0.5724         -0.5724         20681          8843         20678          8842
[slide_hash/sse2 vs. slide_hash/sse2]/16384                       -0.5724         -0.5724         20659          8833         20658          8833
[slide_hash/sse2 vs. slide_hash/sse2]/16384                       -0.5724         -0.5724         20652          8831         20652          8831
[slide_hash/sse2 vs. slide_hash/sse2]/16384                       -0.5733         -0.5732         20660          8815         20654          8815
[slide_hash/sse2 vs. slide_hash/sse2]/16384                       -0.5739         -0.5739         20679          8811         20678          8811
[slide_hash/sse2 vs. slide_hash/sse2]/16384                       -0.5728         -0.5728         20666          8829         20666          8829
[slide_hash/sse2 vs. slide_hash/sse2]/16384_pvalue                 0.0002          0.0002      U Test, Repetitions: 10 vs 10
[slide_hash/sse2 vs. slide_hash/sse2]/16384_mean                  -0.5729         -0.5729         20664          8825         20662          8825
[slide_hash/sse2 vs. slide_hash/sse2]/16384_median                -0.5726         -0.5726         20660          8830         20657          8829
[slide_hash/sse2 vs. slide_hash/sse2]/16384_stddev                +0.0533         +0.0883            11            12            10            11
[slide_hash/sse2 vs. slide_hash/sse2]/16384_cv                    +1.4663         +1.5481             0             0             0             0
[slide_hash/sse2 vs. slide_hash/sse2]/32768                       -0.5738         -0.5738         24829         10581         24828         10581
[slide_hash/sse2 vs. slide_hash/sse2]/32768                       -0.5739         -0.5738         24815         10574         24808         10574
[slide_hash/sse2 vs. slide_hash/sse2]/32768                       -0.5743         -0.5743         24839         10575         24838         10575
[slide_hash/sse2 vs. slide_hash/sse2]/32768                       -0.5732         -0.5732         24810         10588         24806         10588
[slide_hash/sse2 vs. slide_hash/sse2]/32768                       -0.5745         -0.5745         24829         10564         24829         10564
[slide_hash/sse2 vs. slide_hash/sse2]/32768                       -0.5747         -0.5747         24855         10572         24854         10572
[slide_hash/sse2 vs. slide_hash/sse2]/32768                       -0.5759         -0.5759         24894         10558         24894         10558
[slide_hash/sse2 vs. slide_hash/sse2]/32768                       -0.5748         -0.5748         24851         10566         24851         10566
[slide_hash/sse2 vs. slide_hash/sse2]/32768                       -0.5742         -0.5742         24813         10564         24813         10564
[slide_hash/sse2 vs. slide_hash/sse2]/32768                       -0.5734         -0.5734         24827         10591         24827         10590
[slide_hash/sse2 vs. slide_hash/sse2]/32768_pvalue                 0.0002          0.0002      U Test, Repetitions: 10 vs 10
[slide_hash/sse2 vs. slide_hash/sse2]/32768_mean                  -0.5743         -0.5743         24836         10573         24835         10573
[slide_hash/sse2 vs. slide_hash/sse2]/32768_median                -0.5742         -0.5742         24829         10573         24829         10573
[slide_hash/sse2 vs. slide_hash/sse2]/32768_stddev                -0.5743         -0.5955            25            11            26            11
[slide_hash/sse2 vs. slide_hash/sse2]/32768_cv                    +0.0001         -0.0499             0             0             0             0
OVERALL_GEOMEAN                                                   -0.5731         -0.5731             0             0             0             0
```

File: arch/x86/slide_hash_sse2.c [hunk 1]
```diff
@@ -15,18 +15,36 @@
...
     table += entries;
-    table -= 8;
+    table -= 16;
+    __m128i value0, value1, result0, result1;
 
-    do {
-        __m128i value, result;
+    /* ZALLOC allocates this pointer unless the user chose a custom allocator.
+     * Our alloc function is aligned to 64 byte boundaries */
+    if (UNLIKELY(((uintptr_t)table & 15) != 0)) {
```
Comment (nmoinvaz) [inline arch/x86/slide_hash_sse2.c]: @KungFuJesus am I correct in assuming that if #1096 gets merged, then we don't need to account for custom allocators, and we wouldn't then need this branch?

Comment (KungFuJesus) [reply arch/x86/slide_hash_sse2.c]: In theory, yes.  We could then cut the size of the function in half, basically.  This assumes that there's no weird corner cases that could end up messing with it.  Perhaps a debug or runtime assertion at that point would be prudent?  Something to catch potential future misuse so that the user knows the inputs need to be aligned.  An aligned load on an unaligned address, even on an architecture that made this issue virtually disappear, will segfault.

Now, MSVC for some reason completely removes the ability to shoot yourself in the foot (all loads become unaligned when compiling on newer targets).  Probably a blessing and a curse, I imagine.

Comment (KungFuJesus) [reply arch/x86/slide_hash_sse2.c]: At present, anyway, slide_hash is only ever called on members allocated with the ZALLOC macro, so, I could remove the if entirely, sub in a possible sanity check at the beginning of the function (or least just a debug level assertion), and say this commit depends on #1096 ?

I see LONGEST_MATCH contains some Assert() function, should I be using that?  Does this disappear while in debug mode?  I don't expect a user to encounter this issue as these things are only ever allocated by us, so a debug level assert is probably appropriate here.

Comment (KungFuJesus) [reply arch/x86/slide_hash_sse2.c]: With regard to shortening the length of the compiled code: to be honest I'm a little surprised that the compiler, after it does this inlining, isn't smart enough to identify this loop is just repeated.  It would be trivial to swap a pointer and a length variable and jump to the label containing the loop.  Maybe I should try that to see if there's any discernible improvement.  At the very least it'd reduce L1 cache pollution.

Comment (KungFuJesus) [reply arch/x86/slide_hash_sse2.c]: Yeah, no appreciable difference to runtime, just significantly smaller code size.

**Comment by KungFuJesus** [comment]
So, this has an indirect soft dependency on #1096.  In practice, it's not likely to occur, but for the weird cases where it does, such as in node.js, we should get that merged before this.

**Comment by KungFuJesus** [comment]
Resulting code size from that last push:
```
       │    Disassembly of section .text:                                                                                                                                                                                                                                                                                                                                   
       │                                                                                                                                                                                                                                                                                                                                                                    
       │    00000000004057e0 <slide_hash_sse2>:                                                                                                                                                                                                                                                                                                                             
       │    slide_hash_sse2():                                                                                                                                                                                                                                                                                                                                              
  0.05 │      mov     0x40(%rdi),%eax                                                                                                                                                                                                                                                                                                                                       
       │      mov     $0x2,%esi                                                                                                                                                                                                                                                                                                                                             
       │      mov     $0x10000,%edx                                                                                                                                                                                                                                                                                                                                         
  0.01 │      mov     0x60(%rdi),%r8                                                                                                                                                                                                                                                                                                                                        
       │      movd    %eax,%xmm2                                                                                                                                                                                                                                                                                                                                            
       │      movzwl  %ax,%ecx                                                                                                                                                                                                                                                                                                                                              
       │      mov     0x68(%rdi),%rax                                                                                                                                                                                                                                                                                                                                       
       │      pshufb  0x2fdb(%rip),%xmm2        # 4087e0 <crc_table+0x2020>                                                                                                                                                                                                                                                                                                 
       │25:   mov     %edx,%edi                                                                                                                                                                                                                                                                                                                                             
  0.02 │      lea     -0x20(%rax,%rdi,2),%rax                                                                                                                                                                                                                                                                                                                               
       │      nop                                                                                                                                                                                                                                                                                                                                                           
 22.25 │30:   movdqa  (%rax),%xmm1                                                                                                                                                                                                                                                                                                                                          
       │      sub     $0x20,%rax                                                                                                                                                                                                                                                                                                                                            
 20.22 │      movdqa  0x30(%rax),%xmm0                                                                                                                                                                                                                                                                                                                                      
 13.89 │      psubusw %xmm2,%xmm1                                                                                                                                                                                                                                                                                                                                           
  5.29 │      psubusw %xmm2,%xmm0                                                                                                                                                                                                                                                                                                                                           
  1.44 │      movaps  %xmm1,0x20(%rax)                                                                                                                                                                                                                                                                                                                                      
  5.65 │      movaps  %xmm0,0x30(%rax)                                                                                                                                                                                                                                                                                                                                      
 12.12 │      sub     $0x10,%edx                                                                                                                                                                                                                                                                                                                                            
 18.87 │    ↑ jne     30                                                                                                                                                                                                                                                                                                                                                    
  0.06 │      cmp     $0x1,%esi                                                                                                                                                                                                                                                                                                                                             
       │    ↓ jne     60                                                                                                                                                                                                                                                                                                                                                    
  0.10 │    ← ret                                                                                                                                                                                                                                                                                                                                                           
       │      nop                                                                                                                                                                                                                                                                                                                                                           
  0.02 │60:   mov     $0x1,%esi                                                                                                                                                                                                                                                                                                                                             
  0.01 │      mov     %r8,%rax                                                                                                                                                                                                                                                                                                                                              
  0.01 │      mov     %ecx,%edx                                                                                                                                                                                                                                                                                                                                             
  0.01 │    ↑ jmp     25 
```

File: arch/x86/slide_hash_sse2.c [hunk 2]
```diff
@@ -12,27 +12,50 @@
...
+        _mm_store_si128((__m128i *)table, result0);
+        _mm_store_si128((__m128i *)(table + 8), result1);
 
-        table -= 8;
-        entries -= 8;
+        table -= 16;
+        entries -= 16;
     } while (entries > 0);
+
+    ++onChain;
```
Comment (nmoinvaz) [inline arch/x86/slide_hash_sse2.c]: Camel case variables..
Code looks pretty clean otherwise.

Comment (KungFuJesus) [reply arch/x86/slide_hash_sse2.c]: Going to need to write a git hook for this or something when operating on this repo.

**Comment by KungFuJesus** [comment]
Ugh, somehow some way git managed to blow away my SSE4 removal changes.  Man I hate revising revision history :-/.
Err no, looking at the wrong PR here.  I'm a bit puzzled why this doesn't like the slidehash changes.

**Comment by KungFuJesus** [comment]
Ahhh, a lack of alignment.  I guess, do I wait on PR #1096 or do I write code that works agnostic of alignment and axe the second half later?

**Comment by nmoinvaz** [comment]
I always reorder my commits so that the ones least likely to change are first in the history and the the most likely to change are the last.

**Comment by KungFuJesus** [comment]
@nmoinvaz did you see my comments about posix_memalign on #1096 ?

**Comment by KungFuJesus** [comment]
So the Mac OS CI tests are failing for lack of alignment - that much I get, but what's up with the Ubuntu ones?  I didn't touch CMakeLists.txt or configure, it shouldn't be modifying any link time behavior.

Hah, figured it out.  It's the debug assertion itself.  With CMakeLists.txt, they automatically add -DNDEBUG, apparently we don't with configure?  I guess I can add a fix to make this consistent.

**Comment by KungFuJesus** [comment]
Also I think I just verified that this statement:
https://github.com/zlib-ng/zlib-ng/blob/develop/zutil_p.h#L29
is absolutely false.  We should fallback to posix_memalign or _mm_malloc() for macos.  Though Mac OS's man page seems to think it does:
```
The malloc(), calloc(), valloc(), realloc(), and reallocf() functions allocate memory.  The allocated memory is aligned such that it can be used for any data type, including AltiVec- and SSE-related types.
```
For one, it'd be the first realloc in history I've seen that aligns things, but for two, we're throwing an assertion on this pointer, right out of malloc.  Perhaps Apple hasn't updated their man pages in some time?

File: configure [hunk 3]
```diff
@@ -233,7 +233,7 @@ show()
   case "$*" in
     *$test.c*)
       echo "=== $test.c ===" >> configure.log
-      cat $test.c >> configure.log
+     cat $test.c >> configure.log
```
Comment (nmoinvaz) [inline configure]: Looks like whitespace change?

Comment (KungFuJesus) [reply configure]: Hmm, not sure how that managed to sneak in.  Easy fix, though.

File: configure [hunk 4]
```diff
@@ -396,6 +396,9 @@ if test "$gcc" -eq 1 && ($cc $CFLAGS -c $test.c) >> configure.log 2>&1; then
   if test $debug -eq 1; then
     CFLAGS="${CFLAGS} -DZLIB_DEBUG"
     SFLAGS="${SFLAGS} -DZLIB_DEBUG"
+  else
+    CFLAGS="${CFLAGS} -DNDEBUG"
+    SFLAGS="${SFLAGS} -DNDEBUG"
```
Comment (nmoinvaz) [inline configure]: I think perhaps this might be good as a separate PR. It will get in quicker.

**Comment by KungFuJesus** [comment]
This PR now also depends on #1145

**Comment by KungFuJesus** [comment]
Ahh, at second glance the compiler just doesn't like my positioning of parenthesis.  The error on Mac OS is the same on Ubuntu - inconsistent flags between configure and cmake.

**Comment by nmoinvaz** [comment]
@KungFuJesus maybe rebase for CI?

**Comment by KungFuJesus** [comment]
> @KungFuJesus maybe rebase for CI?

Sure.  Is the google test code ready or is there still some weirdness going on with the coverage test?  If we're good to go on that I could just wait for that one before rebasing?  This _kinda sorta_ depends on your force aligned allocation code, which kind of but not really depends on you writing a unit test for it.

**Comment by nmoinvaz** [comment]
@KungFuJesus sorry I'm holding things up on that front. But that PR should prevent this PR from being merged in my opinion, because the bug with non-aligned memory buffers already exists in the current code base. We currently make the assumption that memory is always aligned and so does this code.

**Comment by KungFuJesus** [comment]
Hah, well it does _now_, but yes I made that decision to prevent a future PR from having to occur that just removed the alignment check and branch.  No worries, the people this helps most are those on 11+ish year old hardware, anyway (well, post nehalem pre haswell it helps, _slightly_).  I've got my patched version running locally so there's no rush :).

**Comment by nmoinvaz** [comment]
@KungFuJesus what do you use zlib-ng for? We use it at work as part of our native application as just zlib replacement. It handles all our zlib related calls. Zipping, cURL, web server stuff, etc.

**Comment by KungFuJesus** [comment]
> @KungFuJesus what do you use zlib-ng for? We use it at work as part of our native application as just zlib replacement. It handles all our zlib related calls. Zipping, cURL, web server stuff, etc.

Basically everything I use zlib for, hah.  Mainly use it to speed up web browsing rather than hosting.  If you LD_PRELOAD zlib-ng in front of something like Chrome, page load latency can be night and day, especially if it's riddled with PNG images.  At work I used Intel's zlib patches to speed up PNG exports for a web application server we have written in C# using .NET core in Linux.  We are predominantly using that bit to export palette PNGs so honestly, Intel's custom deflate dictionary + implementation at level 4 _probably_ isn't buying us nearly as much as a faster adler and crc32 checksum might (and as I've demo'd, IPP's routines are slower), but I haven't had the funding on that yet to try such a thing.  

Additionally, I have an ancient mythtv system that I maintain at home and I find the interface on a Core 2 Quad Q9650 is a lot slower than comfortable sometimes, so I've found that injecting zlib-ng there breathes new life into it.  It's mainly burning excess cycles in thumbnail generation, which I suspect it's doing more than it should (I've been meaning to look into that, though).

At work I help maintain a radar predictor that uses HDF5 as intermediate representation.  That too, uses deflate compression under the hood (with zlib routines) and is immensely, painfully slow (especially on non-x86 hardware).  I may end up pulling in zlib-ng on that too, even though it's a one time cost at parse time.  

Oh and one more application I used it for recently - I injected it into our x2goserver services on a few systems.  Latency was astonishingly better (uses zlib for a few things, but mainly tiled PNG compression).

File: arch/x86/slide_hash_sse2.c [hunk 5]
```diff
@@ -12,27 +12,50 @@
...
+
+next_chain:
+    table = (on_chain) ? table1 : table0;
+    entries = (on_chain) ? entries1 : entries0;
 
-static inline void slide_hash_chain(Pos *table, uint32_t entries, const __m128i wsize) {
     table += entries;
-    table -= 8;
+    table -= 16;
+    __m128i value0, value1, result0, result1;
```
Comment (mtl1979) [inline arch/x86/slide_hash_sse2.c]: All variables should be declared before any other statements in current scope.

**Comment by Dead2** [comment]
Xeon E5-2650, x86-64, GCC 4.8
 
## Baseline ab6665b1bee170c27d229899ed23e6053ebd900b GCC4.8 SSE2-only
```
 Tool: minideflate Levels: 0-9
 Runs: 40         Trim worst: 20

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.015%      0.010/0.019/0.023/0.003        0.016/0.024/0.029/0.003       15,738,731
 1     54.177%      0.242/0.259/0.268/0.007        0.121/0.131/0.138/0.005        8,525,423
 2     43.871%      0.447/0.460/0.466/0.005        0.131/0.138/0.142/0.003        6,903,690
 3     42.387%      0.564/0.580/0.588/0.006        0.122/0.134/0.139/0.004        6,670,227
 4     41.647%      0.614/0.635/0.645/0.008        0.121/0.129/0.134/0.005        6,553,734
 5     41.225%      0.686/0.704/0.712/0.007        0.106/0.126/0.132/0.007        6,487,311
 6     41.043%      0.767/0.786/0.794/0.006        0.114/0.126/0.134/0.005        6,458,729
 7     40.778%      0.974/0.992/1.000/0.007        0.119/0.127/0.132/0.004        6,416,929
 8     40.704%      1.187/1.206/1.218/0.008        0.111/0.125/0.129/0.005        6,405,237
 9     40.409%      1.330/1.346/1.355/0.008        0.120/0.129/0.134/0.005        6,358,939

 avg1  48.626%                        0.699                          0.119
 avg2  54.029%                        0.776                          0.132
 tot                                139.721                         23.752       76,518,950

   text    data     bss     dec     hex filename
 116535    1456      48  118039   1cd17 libz-ng.so.2
``` 
## PR #1141 4bd884207a3467e29dde8f1197d7f46ff4436d87 GCC4.8 SSE2-only
```
 Tool: minideflate Levels: 0-9
 Runs: 40         Trim worst: 20

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.015%      0.013/0.018/0.023/0.003        0.017/0.022/0.027/0.003       15,738,731
 1     54.177%      0.247/0.262/0.268/0.006        0.122/0.132/0.138/0.005        8,525,423
 2     43.871%      0.444/0.459/0.465/0.006        0.131/0.138/0.144/0.004        6,903,690
 3     42.387%      0.559/0.580/0.588/0.008        0.118/0.131/0.138/0.006        6,670,227
 4     41.647%      0.607/0.630/0.638/0.007        0.120/0.132/0.136/0.004        6,553,734
 5     41.225%      0.682/0.698/0.707/0.007        0.109/0.123/0.132/0.007        6,487,311
 6     41.043%      0.772/0.784/0.791/0.005        0.117/0.127/0.131/0.004        6,458,729
 7     40.778%      0.973/0.988/0.998/0.009        0.116/0.126/0.132/0.005        6,416,929
 8     40.704%      1.188/1.204/1.213/0.008        0.108/0.122/0.131/0.007        6,405,237
 9     40.409%      1.326/1.343/1.352/0.006        0.116/0.127/0.133/0.005        6,358,939

 avg1  48.626%                        0.697                          0.118
 avg2  54.029%                        0.774                          0.131
 tot                                139.319                         23.611       76,518,950

   text    data     bss     dec     hex filename
 116535    1456      48  118039   1cd17 libz-ng.so.2
```

Performance is too close to tell whether it is really faster, but at least there does not seem to be any regressions.

**Comment by KungFuJesus** [comment]
> Performance is too close to tell whether it is really faster, but at least there does not seem to be any regressions.

As the commit message explicitly states, this only makes a significant difference on pre-nehalem hardware (Core 2 era).  You are going to see at best a tiny bit of a gain from the unroll that happens.

**Review by Dead2 — APPROVED** [review]

**Comment by nmoinvaz** [comment]
This shouldn't be dependent on the other PR #1096. Because it is only when using custom allocator, and it is not the job of this code to compensate for that. Also if they are using a custom allocator with non 64 byte alignment they will likely run into other problems.
